[quant][graphmode][fx] Attach a weight qparam dict to linear and conv in reference quantized model

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62861
* #62698
* #62608
* #62607
* #62489
* __->__ #62488

Summary:
Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

TODO: the numerics for linear and conv in reference quantized model is still not correct since
we did not quantize weight, we may explore things like parameterization to implement this support

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30053979](https://our.internmc.facebook.com/intern/diff/D30053979)

jerryzh168 committed:
[quant][graphmode][fx] Attach a weight qparam dict to linear and conv in reference quantized model

Summary:
Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

TODO: the numerics for linear and conv in reference quantized model is still not correct since
we did not quantize weight, we may explore things like parameterization to implement this support

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Attach a weight qparam dict to linear and conv in reference quantized model"


Summary:
Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

TODO: the numerics for linear and conv in reference quantized model is still not correct since
we did not quantize weight, we may explore things like parameterization to implement this support

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Attach a weight qparam dict to linear and conv in reference quantized model"


Summary:
Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

TODO: the numerics for linear and conv in reference quantized model is still not correct since
we did not quantize weight, we may explore things like parameterization to implement this support

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30053979).

vkuzo commented:
> Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

Are we sure that the additional complexity is worth the benefits?  I don't feel strongly one way or the other, just surprised to see model size as something to optimize for for a reference model.

jerryzh168 commented:
> > Instead of attaching weight observer/fake_quant to the float linear and conv, we can
> > compute the quantization parameters and attach that as a dictionary to these modules so
> > that we can reduce the model size and make the reference module clearer
> 
> Are we sure that the additional complexity is worth the benefits? I don't feel strongly one way or the other, just surprised to see model size as something to optimize for for a reference model.

oh, it's both model size and clarity, otherwise we need to call calculate_qparams etc. I think it's more intuitive to attach the qparams directly. what do you think?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Attach a weight qparam dict to linear and conv in reference quantized model"


Summary:
Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

TODO: the numerics for linear and conv in reference quantized model is still not correct since
we did not quantize weight, we may explore things like parameterization to implement this support

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30053979](https://our.internmc.facebook.com/intern/diff/D30053979)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30053979).

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -597,9 +600,10 @@ def convert(self,
                     # run weight observer
-                    float_conv.weight_post_process(float_conv.weight)
+                    weight_post_process(float_conv.weight)
+                float_conv.weight_qparams = get_qparam_dict(weight_post_process)
```
is there an example how this will be used?  I see it set, but not how it is queried.

I think the additional complexity will depend on the callsite and how it does serialization.

jerryzh168 replied:
It will be used in the backend lowering pass to generate quantize node for weight, I don't have example code currently. basically the weight_qparams dictionary will be: {'qschem': ..., 'scale': ..., 'zero_point': ..., 'dtype': ...., ....} and backend lowering pass can use this info to quantize the weight.

I don't think this is additional complexity, why do you think this is additional complexity?

vkuzo replied:
is there going to be a need for custom versioned setstate/getstate functions to hold this information?

jerryzh168 replied:
oh I see, yeah probably, I haven't verified that. I think it's still worth it though since otherwise the backend lowering pass would need to extrac t this info, this also exposes less implementation detail to the backend lowering pass as well

jerryzh168 committed:
Update on "[quant][graphmode][fx] Attach a weight qparam dict to linear and conv in reference quantized model"


Summary:
Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

TODO: the numerics for linear and conv in reference quantized model is still not correct since
we did not quantize weight, we may explore things like parameterization to implement this support

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30053979](https://our.internmc.facebook.com/intern/diff/D30053979)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30053979).

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -597,9 +618,12 @@ def convert(self,
+                    weight_post_process(float_conv.weight)
+                float_conv._weight_qparams = get_qparam_dict(weight_post_process)
+                float_conv._register_state_dict_hook(_save_weight_qparams)
+                float_conv._register_load_state_dict_pre_hook(_load_weight_qparams, with_module=True)
```
Have we considered having a `ReferenceConvNd` module instead, so these things would be defined on the class and easy to discover?  I could see a user printing out their model, seeing `nn.Conv2d` in the logs, not seeing any of these extra information in the class definition, and being really confused on where they came from.

jerryzh168 replied:
Agree that having different module would be more clear. I think not introducing new modules have other benefit though, for example simplifying the configurations: now we don't have quantized reference functions, what we need to specify is just operator and the type of operator, this is much simpler than having a mapping as we did before. Having a reference module here would make the configuration more complicated in this case.

This behavior will be clearly documented since it's needed for backend developers to lower the module correctly. If the concern is the logs I think maybe we could change the `_name` or something for the nn.Conv2d?

vkuzo replied:
the concern is that we will have `nn.Conv2d` with logic which is not present in the class definition, which some would consider technical debt.  Are there any ways around this which do not involve adding things to `nn.Conv2d` instances without adding them to the class definition?

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -597,9 +618,12 @@ def convert(self,
                     # run weight observer
-                    float_conv.weight_post_process(float_conv.weight)
+                    weight_post_process(float_conv.weight)
+                float_conv._weight_qparams = get_qparam_dict(weight_post_process)
```
have we considered storing `scale` and `zero_point` as individual buffers, to avoid the need for custom serialization handling?

jerryzh168 replied:
hmm, looks like we can only store Optional[Tensor] in buffers: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L271, it works for `scale` and `zero_point`, but we also need to store qscheme and dtype, and I feel it might be a bit harder?

vkuzo reviewed (approved):
accepting to unblock

jerryzh168 committed:
Update on "[quant][graphmode][fx] Attach a weight qparam dict to linear and conv in reference quantized model"


Summary:
Instead of attaching weight observer/fake_quant to the float linear and conv, we can
compute the quantization parameters and attach that as a dictionary to these modules so
that we can reduce the model size and make the reference module clearer

TODO: the numerics for linear and conv in reference quantized model is still not correct since
we did not quantize weight, we may explore things like parameterization to implement this support

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30053979](https://our.internmc.facebook.com/intern/diff/D30053979)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30053979).
